tests/network/services.go

```diff
@@ -0,0 +1,159 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"fmt"
+	"strconv"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	batchv1 "k8s.io/api/batch/v1"
+	v12 "k8s.io/api/core/v1"
+	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests"
+	cd "kubevirt.io/kubevirt/tests/containerdisk"
+)
+
+var _ = SIGDescribe("Services", func() {
+	var virtClient kubecli.KubevirtClient
+
+	runTCPClientExpectingHelloWorldFromServer := func(host, port, namespace string) *batchv1.Job {
+		job := tests.NewHelloWorldJob(host, port)
+		job, err := virtClient.BatchV1().Jobs(namespace).Create(job)
+		ExpectWithOffset(1, err).ToNot(HaveOccurred())
+		return job
+	}
+
+	BeforeEach(func() {
+		var err error
+		virtClient, err = kubecli.GetKubevirtClient()
+		Expect(err).NotTo(HaveOccurred(), "Should successfully initialize an API client")
+	})
+
+	Context("bridge interface binding", func() {
+		var inboundVMI *v1.VirtualMachineInstance
+		var serviceName string
+
+		const (
+			selectorLabelKey = "expose"
+			selectorLabelValue = "me"
+			servicePort = 1500
+		)
+
+		BeforeEach(func() {
+			// inboundVMI expects implicitly to be added to the pod network
+			inboundVMI = tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n")
+			inboundVMI.Labels = map[string]string{selectorLabelKey: selectorLabelValue}
+			inboundVMI.Spec.Subdomain = "myvmi"
+			inboundVMI.Spec.Hostname = "my-subdomain"
```

Comment:
Can we use something more descriptive?

Reply:
What would the advantage be ? I'm porting an existing test, which already has these values.

Updating the chosen values for it seems a bit off-scope.

Reply:
> > Could you please extract the dual-stack distinction logic to wrapper function for creating and deleting services (lines 212-224, 258-266)?
> 
> > Please do the same for the functions assertConnectionToExposedService assertConnectionToExposedServiceFailsat the commitnetwork, tests: use assert / failed connectivity checks`
> 
> I honestly think you're attempting to prematurely optimize this, without having a clear picture of what other tests require (at this moment, I don't think anyone has the clear picture of it).
> 
> What I want, is to have one service on ipv4 based lanes, and 2 services on dual-stacked lanes. Then I want to check that all services have / don't have connectivity.
> 
> Other tests will operate on dual stack on a different way. I don't pretend I know what it will be. In time, _if_ such a pattern arises, we can refactor it.
> 
> I think the test is pretty clear as is, and consistently makes a single question: is the cluster in dual stack mode ? If so, it will:
> 
> * setup test: create 2 services instead of one
> * test: check connectivity to 2 services instead of one
> * teardown: remove 2 services instead of removing one
> 
> I see 0 advantages on encapsulating that on functions - e.g. `testSingleStackService` , `testDualStackService` ...
> 
> What I could do, if you **really** insist, is to push all services to an array, then check connectivity to **all** services in the array, and delete **all** services in the array in the teardown. That would enable me to make 'the' question a single time, and represents better what the tests do.
> 
> > It will keep the flow of the code and make it more clean,
> > Having such encapsulate make it easier to maintain and extend (for example; in case we will need to do the same for single stack IPv6 )
> 
> AFAIK, single stack IPv6 is simply not a thing we'll ever support.
> 
> The encapsulation is already done: I have a 'connectivity checker' and I point it at a service name + port. Then I expect it to work / fail. Depending on the use case at hand (dual-stack) I might need to point at more than one thing.
`
I understand what you did and I think its great that you moved it to another file 
and exported most of the test code to functions :)

Dont get me wrong I am not trying to optimize something we dont fully understand  yet
nor search for a silver built to all tests adapting to dual stack.

In this test I saw a classic case we can improve by encapsulating "distingush cluster is dual-stack logic".
The big advantage is that we wont disrupt the flow of the test code with `if` statements, 
making it easy to follow and understand.
Is also make the code easy to maintain if we will need to change anything (IPv6 stack was just an example I know its not a thing right now)

I am not sure how using services array will encapsulate that mentioned logic,
I think you will have to check if cluster is dual stack to add the ipv6 service to the array.

Reply:
I commented here by accident, lets continue [here](https://github.com/kubevirt/kubevirt/pull/4035#issuecomment-681845972)

Reply:
inbound.vmi looks nicer than myvmi.my-subdomain.
Same arguments may led us to see myVMI instead inboundVMI and so on...

Sorry if I take it to an extreme, but these little things hurt my eyes. If it bothers you too much, do not change it.